### PR TITLE
23092 - Cont in Business Only Alter to Cont in Type

### PR DESCRIPTION
**Pull request by ArwenQin**
*Issue #:* /bcgov/entity#23092

*Description of changes:*
- A C company can alter to a C Type (Cont In ULC, C, CCC, BEN)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-edit-dev--pr-590-mkdch8go.web.app

**Comment by ArwenQin** [comment]
Now a Cont in business can alter to other valid Cont in legal types successfully.
Test with:
Cont In ULC to Cont In Benefit or Limited
https://business-edit-dev--pr-590-mkdch8go.web.app/C9900774/alteration?accountid=3040

Cont In Limited to Cont In BEN, ULC, or CCC
https://business-edit-dev--pr-590-mkdch8go.web.app/C9900734/alteration?accountid=3040

File: src/components/common/YourCompany/ChangeBusinessType.vue [hunk 1]
```diff
@@ -458,6 +458,16 @@ export default class ChangeBusinessType extends Mixins(CommonMixin) {
...
+    // if the original legal type is one of the continued-in types, alter to another Cont. In type
+    if (
+      this.getOriginalLegalType === CorpTypeCd.BEN_CONTINUE_IN ||
+      this.getOriginalLegalType === CorpTypeCd.CCC_CONTINUE_IN ||
+      this.getOriginalLegalType === CorpTypeCd.CONTINUE_IN ||
+      this.getOriginalLegalType === CorpTypeCd.ULC_CONTINUE_IN
+    ) {
+      this.selectedEntityType = convertToContinuedInEntityType(this.selectedEntityType)
+    }
+
```
Comment (severinbeauvais) [inline src/components/common/YourCompany/ChangeBusinessType.vue:470]: I'm thinking you can remove `convertToContinuedInEntityType()` ...?

Comment (severinbeauvais) [reply src/components/common/YourCompany/ChangeBusinessType.vue:470]: See also `convertToRegularEntityType()`.

Ref: https://github.com/bcgov/business-edit-ui/pull/582/files#diff-ca058487aa6730e697018454f9201fad0a0c649b2084211046382fa3e93b6c61

Comment (ArwenQin) [reply src/components/common/YourCompany/ChangeBusinessType.vue:470]: Are these 2 methods used in the previous ticket to solve the issue that for Cont in, if the user clicks on "Change" for the Business Type then the Business Type Selection is blank?

**Comment by severinbeauvais** [comment]
I cannot test the preview links because Auth API is not returning settings or authorizations.

File: src/components/common/YourCompany/ChangeBusinessType.vue (hunk 1 above)
Comment (severinbeauvais) [reply src/components/common/YourCompany/ChangeBusinessType.vue:470]: Yes..

**Comment by ArwenQin** [comment]
> [https://business-edit-dev--pr-590-mkdch8go.web.app/C9900734/](https://business-edit-dev--pr-590-mkdch8go.web.app/C9900734/alteration?accountid=3040)

Oh yes I see. Could you please go to the dashboard of C9900734 and C9900734, and try to alter them?

**Comment by severinbeauvais** [comment]
> > I cannot test the preview links because Auth API is not returning settings or authorizations.
> 
> Oh yes I see... I tried, not working either

I will checkout your code and try it locally. (`gh pr checkout 590`)

**Comment by severinbeauvais** [comment]
OK, I tested locally the first business (CUL -> CBEN) and the filing JSON looks correct (see below), but I still think there may be an easier solution.

image

**Comment by ArwenQin** [comment]
> OK, I tested locally the first business (C9900774) and the filing JSON looks correct (see below), but I still think there may be an easier solution.
> 
> image

Hi Sev, I just updated the solution. I also corrected the entityTypeOptions property for C, CBEN and CUL

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-edit-dev--pr-590-mkdch8go.web.app

File: src/components/common/YourCompany/ChangeBusinessType.vue [hunk 2]
```diff
@@ -402,7 +391,7 @@ export default class ChangeBusinessType extends Mixins(CommonMixin) {
     if (this.isNumberedCompany) {
       return false
     }
-    // Named companies to CC or ULC require a name request.
+    // Named companies to CC/CCC or ULC require a name request.
```
Comment (severinbeauvais) [inline src/components/common/YourCompany/ChangeBusinessType.vue:394]: Also CUL I think.

File: src/utils/resource-utils.ts [hunk 3]
```diff
@@ -2,7 +2,11 @@ import {
   AlterationResourceBc,
   AlterationResourceBen,
   AlterationResourceCc,
-  AlterationResourceUlc
+  AlterationResourceUlc,
+  AlterationResourceCul,
+  AlterationResourceC,
+  AlterationResourceCben,
+  AlterationResourceCcc
```
Comment (severinbeauvais) [inline src/utils/resource-utils.ts:9]: Please alphabetize these (and the object below).

File: src/components/common/YourCompany/ChangeBusinessType.vue [hunk 4]
```diff
@@ -503,22 +473,26 @@ export default class ChangeBusinessType extends Mixins(CommonMixin) {
...
-  /** Whether current entity selection is a BC ULC Company. */
+  /** Whether current entity selection is a ULC/CUL Company. */
   get isUnlimitedLiability (): boolean {
-    return (this.selectedEntityType === CorpTypeCd.BC_ULC_COMPANY)
+    return (this.selectedEntityType === CorpTypeCd.BC_ULC_COMPANY ||
+    this.selectedEntityType === CorpTypeCd.ULC_CONTINUE_IN)
   }
 
-  /** Whether current entity selection is a BC CCC. */
+  /** Whether current entity selection is a CCC. */
```
Comment (severinbeauvais) [inline src/components/common/YourCompany/ChangeBusinessType.vue:486]: Please be consistent with the comments on lines 474, 480 and 486.

Comment (severinbeauvais) [reply src/components/common/YourCompany/ChangeBusinessType.vue:486]: (What I'd do is match the comments to the method name.)

**Comment by severinbeauvais** [comment]
> Hi Sev, I just updated the solution. I also corrected the entityTypeOptions property for C, CBEN and CUL

OK, this "feels like" the solution I was expecting.

However, I have a done a bit of testing and there are issues (though maybe not caused by this code)...

1. C9900774 is, at the moment, of type "C". But if try to create a name-change NR for it, business search (from Namerequest UI) says it's type CUL. Please check with Vysakh whether this should have been synced.

2. Here is the draft alteration filing (where I changed type from C to CBEN). Do you know if filing.alteration.nameRequest.legalType should also be CBEN? (Maybe ask Vysakh again.)

image

Please test thoroughly, including with new NR to change the business name. Thanks.

**Comment by ArwenQin** [comment]
Thanks Sev. I just updated the comments/order.
I will check with Vysakh

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-edit-dev--pr-590-mkdch8go.web.app

**Comment by ArwenQin** [comment]
@severinbeauvais 
I just did thorough tests, the type alter works well. Only the Namerequest type pending Vysakh's confirmation.
By the way, I found another bug in the 3. Court Order and Plan of Arrangement section:
image
If I enter the court order number only, and don't check the POA box, there shows no error; but when we file it, there is a schema error.
If we don't enter the court order number, or enter court order number & check the box - all good.
Just to confirm, what is the requirement? Is court order number only + no check box allowed?
Shall I create a new ticket for this?

**Comment by severinbeauvais** [comment]
> Just to confirm, what is the requirement? Is court order number only + no check box allowed? Shall I create a new ticket for this?

It should be possible to enter a court number and not check the POA box. This is probably a simple fix you can do in this ticket. According to the [court order schema](https://github.com/bcgov/business-schemas/blob/main/src/registry_schemas/schemas/court_order.json), `effectOfOrder` is a string but isn't nullable, so you probably need to remove that property if it's empty.

Please test a ChangeOfRegistration ("firm alteration") filing for the same bug, and fix that, too.

If you really want, you can create a new ticket for this. (And then just work on it today.)

File: src/mixins/filing-template-mixin.ts [hunk 5]
```diff
@@ -331,7 +332,7 @@ export default class FilingTemplateMixin extends DateMixin {
     if (this.getHasPlanOfArrangement || this.getFileNumber) {
       filing.alteration.courtOrder = {
         fileNumber: this.getFileNumber,
-        effectOfOrder: this.getHasPlanOfArrangement ? EffectOfOrders.PLAN_OF_ARRANGEMENT : null,
+        effectOfOrder: this.getHasPlanOfArrangement ? EffectOfOrders.PLAN_OF_ARRANGEMENT : undefined,
         hasPlanOfArrangement: this.getHasPlanOfArrangement
```
Comment (ArwenQin) [inline src/mixins/filing-template-mixin.ts:336]: corrected the EffectOfOrders schema for Court Order section

File: src/mixins/filing-template-mixin.ts [hunk 6]
```diff
@@ -304,6 +304,7 @@ export default class FilingTemplateMixin extends DateMixin {
         // add in the fields needed by Legal API
         ...this.getNameRequest,
         legalName: this.getNameRequestLegalName || undefined, // don't include if empty
+        legalType: this.getEntityType,
         nrNumber: this.getNameRequestNumber
```
Comment (ArwenQin) [inline src/mixins/filing-template-mixin.ts:308]: image
Per Vysakh's notes, I updated the lteration.nameRequest.legalType to be synced with the altered legal type

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-edit-dev--pr-590-mkdch8go.web.app

File: src/mixins/filing-template-mixin.ts (hunk 6 above)
Comment (ArwenQin) [reply src/mixins/filing-template-mixin.ts:308]: But for the first issue, I tested with other businesses, the alterations are not synced to colin either. I'm not familiar with this, may I know the possible reasons?

Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:308]: Vysakh identified that the sync didn't work for some reason. I just wanted him to know about it. That's outside the scope of this ticket so don't worry about it.

File: src/mixins/filing-template-mixin.ts [hunk 7]
```diff
@@ -209,7 +209,7 @@ export default class FilingTemplateMixin extends DateMixin {
       if (this.getHasPlanOfArrangement || this.getFileNumber) {
         filing.correction.courtOrder = {
           fileNumber: this.getFileNumber,
-          effectOfOrder: this.getHasPlanOfArrangement ? EffectOfOrders.PLAN_OF_ARRANGEMENT : null,
+          effectOfOrder: this.getHasPlanOfArrangement ? EffectOfOrders.PLAN_OF_ARRANGEMENT : undefined,
```
Comment (severinbeauvais) [inline src/mixins/filing-template-mixin.ts:212]: Please verify by filing a correction with a court order number and with / without a POA.

File: src/mixins/filing-template-mixin.ts [hunk 8]
```diff
@@ -647,7 +648,7 @@ export default class FilingTemplateMixin extends DateMixin {
     if (this.getHasPlanOfArrangement || this.getFileNumber) {
       filing.changeOfRegistration.courtOrder = {
         fileNumber: this.getFileNumber,
-        effectOfOrder: this.getHasPlanOfArrangement ? EffectOfOrders.PLAN_OF_ARRANGEMENT : null,
+        effectOfOrder: this.getHasPlanOfArrangement ? EffectOfOrders.PLAN_OF_ARRANGEMENT : undefined,
```
Comment (severinbeauvais) [inline src/mixins/filing-template-mixin.ts:651]: Please verify by filing a change of registration with a court order number and with / without a POA.

**Review by severinbeauvais — APPROVED** [review]
The code looks good but please ensure you have tested these scenarios before merging (and also add Test Notes in the ticket to let Riyaz know what you've verified locally):
- BC to another type
- BEN to another type
- CC to another type
- ULC to another type
- C to another type
- CBEN to another type
- CCC to another type
- CUL to another type
- use name change NRs for the above wherever possible (if you find bugs related to this then create another ticket)

Thanks for the good work 👍

File: src/mixins/filing-template-mixin.ts (hunk 7 above)
Comment (ArwenQin) [reply src/mixins/filing-template-mixin.ts:212]: Tested ok with CP1002614

File: src/mixins/filing-template-mixin.ts (hunk 8 above)
Comment (ArwenQin) [reply src/mixins/filing-template-mixin.ts:651]: Tested ok with FM1085277

**Pull request merged**